src/transport_ucx.c

```diff
@@ -0,0 +1,373 @@
+            size_t chunk_idx = 4;
+
+            key[chunk_idx] = '0' + chunk/RUNTIME_ADDR_CHUNK;
+
+            ret = shmem_runtime_get(i, key, addr_bytes+chunk, MIN(len-chunk, RUNTIME_ADDR_CHUNK));
+
+            if (ret) {
+                RAISE_ERROR_MSG("Runtime get of UCX address chunk %zu failed (chunk %d)\n",
+                                chunk/RUNTIME_ADDR_CHUNK, RUNTIME_ADDR_CHUNK);
+            }
+        }
+
+        params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
+        params.address    = shmem_transport_peers[i].addr;
+
+        status = ucp_ep_create(shmem_transport_ucp_worker, &params, &shmem_transport_peers[i].ep);
+        UCX_CHECK_STATUS(status);
+
+        ret = shmem_runtime_get(i, "data_rkey_len", &rkey_len, sizeof(size_t));
+        if (ret) RAISE_ERROR_MSG("Runtime get of UCX data rkey length failed (PE %d, ret %d)\n", i, ret);
+        rkey = malloc(rkey_len);
+        if (rkey == NULL) RAISE_ERROR_MSG("Out of memory, allocating rkey buffer (len = %zu)\n", rkey_len);
+        ret = shmem_runtime_get(i, "data_rkey", rkey, rkey_len);
+        if (ret) RAISE_ERROR_MSG("Runtime get of UCX data rkey failed (PE %d, ret %d)\n", i, ret);
+        status = ucp_ep_rkey_unpack(shmem_transport_peers[i].ep, rkey, &shmem_transport_peers[i].data_rkey);
+        UCX_CHECK_STATUS(status);
+        free(rkey);
+
+        ret = shmem_runtime_get(i, "heap_rkey_len", &rkey_len, sizeof(size_t));
+        if (ret) RAISE_ERROR_MSG("Runtime get of UCX heap rkey length failed (PE %d, ret %d)\n", i, ret);
+        rkey = malloc(rkey_len);
+        if (rkey == NULL) RAISE_ERROR_MSG("Out of memory, allocating rkey buffer (len = %zu)\n", rkey_len);
+        ret = shmem_runtime_get(i, "heap_rkey", rkey, rkey_len);
+        if (ret) RAISE_ERROR_MSG("Runtime get of UCX heap rkey failed (PE %d, ret %d)\n", i, ret);
+        status = ucp_ep_rkey_unpack(shmem_transport_peers[i].ep, rkey, &shmem_transport_peers[i].heap_rkey);
+        UCX_CHECK_STATUS(status);
+        free(rkey);
+
+#ifndef ENABLE_REMOTE_VIRTUAL_ADDRESSING
+        ret = shmem_runtime_get(i, "data_base", &shmem_transport_peers[i].data_base, sizeof(void*));
+        if (ret) RAISE_ERROR_MSG("Runtime get of UCX data base address failed (PE %d, ret %d)\n", i, ret);
+
+        ret = shmem_runtime_get(i, "heap_base", &shmem_transport_peers[i].heap_base, sizeof(void*));
+        if (ret) RAISE_ERROR_MSG("Runtime get of UCX heap base address failed (PE %d, ret %d)\n", i, ret);
+#endif
+    }
+
+    if (shmem_internal_params.PROGRESS_INTERVAL > 0)
+        pthread_create(&shmem_transport_ucx_progress_thread, NULL,
+                       &shmem_transport_ucx_progress_thread_func, NULL);
+
+    return 0;
+}
+
+int shmem_transport_fini(void)
+{
+    ucs_status_t status;
+    int i;
+    void *progress_out;
+
+    __atomic_store_n(&shmem_transport_ucx_progress_thread_enabled, 0, __ATOMIC_RELEASE);
+    pthread_join(shmem_transport_ucx_progress_thread, &progress_out);
```

Comment:
Should we do this when the progress thread is disabled?

Reply:
I don't think it would hurt to check whether the progress thread is enabled before joining.  A quick test I did suggests `pthread_join` returns nonzero if the thread never started.